### DISCUSSION TITLE
Regarding the "In memory node" setup

### DISCUSSION THREAD

**Discussion by rkoranne0755** [zkSync-Community-Hub/zksync-developers#720] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement and the issue you are running into.

![Screenshot from 2024-10-08 18-54-55](https://github.com/user-attachments/assets/7354f83c-12db-4334-84f9-c2d24727b49e)
I'm just following the steps as given in the docs to set-up the "In memory node", but it's showing the error as:

npx zksync-cli dev start

Installing "In memory node"...
WARN[0000] /home/im3/.nvm/versions/node/v18.20.4/lib/node_modules/zksync-cli/node_modules/zkcli-in-memory-node/docker-compose.yml: the attribute `version` is obsolete, it will be ignored, please remove it to avoid potential confusion 
[+] Building 1.8s (9/10)                                                                                                                            docker:desktop-linux
 => [zksync internal] load build definition from Dockerfile                                                                                                         0.0s
 => => transferring dockerfile: 787B                                                                                                                                0.0s
 => [zksync internal] load metadata for docker.io/library/ubuntu:latest                                                                                             1.4s
 => [zksync internal] load .dockerignore                                                                                                                            0.0s
 => => transferring context: 2B                                                                                                                                     0.0s
 => [zksync 1/7] FROM docker.io/library/ubuntu:latest@sha256:b359f1067efa76f37863778f7b6d0e8d911e3ee8efa807ad01fbf5dc1ef9006b                                       0.0s
 => CACHED [zksync 2/7] WORKDIR /app                                                                                                                                0.0s
 => CACHED [zksync 3/7] RUN apt-get update   && apt-get install -y curl jq   && apt-get clean   && rm -rf /var/lib/apt/lists/*                                      0.0s
 => CACHED [zksync 4/7] RUN echo "In memory node version: v0.1.0-alpha.29"                                                                                          0.0s
 => CACHED [zksync 5/7] RUN curl -s echo "https://api.github.com/repos/matter-labs/era-test-node/releases/tags/v0.1.0-alpha.29" |   jq -r '.assets[] | select(.nam  0.0s
 => ERROR [zksync 6/7] RUN tar xz -f era_test_node.tar.gz -C /usr/local/bin/                                                                                        0.3s
------                                                                                                                                                                   
 > [zksync 6/7] RUN tar xz -f era_test_node.tar.gz -C /usr/local/bin/:                                                                                                   
0.222 tar (child): era_test_node.tar.gz: Cannot open: No such file or directory
0.222 tar (child): Error is not recoverable: exiting now
0.223 tar: Child returned status 2
0.223 tar: Error is not recoverable: exiting now
------
failed to solve: process "/bin/sh -c tar xz -f era_test_node.tar.gz -C /usr/local/bin/" did not complete successfully: exit code: 2
ⓘ There was an error while starting the testing environment:
ⓘ Command exited with code 17: 
file:///home/im3/.nvm/versions/node/v18.20.4/lib/node_modules/zksync-cli/bin/utils/helpers.js:52
                reject(new Error(`Command exited with code ${code}: ${errorOutput}`));
                       ^

Error: Command exited with code 17: 
    at ChildProcess.<anonymous> (file:///home/im3/.nvm/versions/node/v18.20.4/lib/node_modules/zksync-cli/bin/utils/helpers.js:52:24)
    at ChildProcess.emit (node:events:517:28)
    at maybeClose (node:internal/child_process:1098:16)
    at ChildProcess._handle.onexit (node:internal/child_process:303:5) {
  level: 'error',
  [Symbol(level)]: 'error',
  [Symbol(message)]: '\x1B[91mⓘ Command exited with code 17: \x1B[39m'
}

Node.js v18.20.4
![Screenshot from 2024-10-08 18-54-55](https://github.com/user-attachments/assets/9c929d42-5973-4b0b-9b71-e33770cc7435)

**Comment by alhenaai** [comment]
It appears that the error you're encountering during the setup of the "In memory node" is related to the extraction of the `era_test_node.tar.gz` file. The error message indicates that the file could not be found, which suggests that the download step might have failed or the file did not download correctly.

Here are a few steps you can take to resolve this issue:

1. **Verify Network and Permissions:**
   - Ensure that your network connection is stable.
   - Check if there are any permissions or firewall settings that might be blocking the download.

2. **Manual Download and Extraction:**
   - Try manually downloading the `era_test_node.tar.gz` from the [latest release](https://github.com/matter-labs/era-test-node/releases/latest) on GitHub.
   - Once downloaded, manually extract it to the desired directory using the command:
     ```bash
     tar xz -f path/to/downloaded/era_test_node.tar.gz -C /usr/local/bin/
     ```

3. **Re-run the Setup:**
   - After manually ensuring that the `era_test_node` is correctly placed, try running the `npx zksync-cli dev start` command again.

4. **Check Docker and Dependencies:**
   - Ensure that Docker is properly installed and running on your system.
   - Verify that all dependencies required by the `zksync-cli` are correctly installed.

If the issue persists after following these steps, it might be helpful to check the system logs or console output for any additional errors that could provide more insight into what might be going wrong.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Reply by rkoranne0755** [reply, marked as answer]
Thank you very much for your help!

I manually ensured that the era_test_node was correctly placed, but the error still persisted. I might have not done it the correct way or something

To resolve the issue, I replaced the "#Download the release version specified by the build argument" section with the following line in the Dockerfile located at /home/#USER/.nvm/versions/node/v20.16.0/lib/node_modules/zksync-cli/node_modules/zkcli-in-memory-node/Dockerfile:

RUN curl -s "https://api.github.com/repos/matter-labs/era-test-node/releases/tags/${LATEST_RELEASE}" | \
  jq -r '.assets[] | select(.name | contains("x86_64-unknown-linux-gnu")) | .browser_download_url' | \
  head -n 1 | xargs -I {} curl -L -o era_test_node.tar.gz {}"
  
The issue was caused by an echo statement that was present, which led to the error.
After making this change, it worked successfully.

Thanks for the help!!!

**Reply by alhenaai** [reply]
You're welcome! I'm glad to hear that you were able to resolve the issue by modifying the Dockerfile to correctly handle the download of the `era_test_node.tar.gz` file. It sounds like the echo statement was indeed causing the problem by not properly handling the download URL.

If you have any more questions or need further assistance with your ZKsync setup or any other issues, feel free to reach out. Happy coding with ZKsync!

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.
